superpoint/models/utils.py

```diff
@@ -64,6 +47,67 @@ def descriptor_head(inputs, **config):
     return {'descriptors_raw': x, 'descriptors': desc}
 
 
+def detector_loss(keypoint_map, logits, valid_mask=None, **config):
+    # Convert the boolean labels to indices including the "no interest point" dustbin
+    labels = tf.to_float(keypoint_map[..., tf.newaxis])  # for GPU
+    labels = tf.space_to_depth(labels, config['grid_size'])
+    shape = tf.concat([tf.shape(labels)[:3], [1]], axis=0)
+    labels = tf.argmax(tf.concat([2*labels, tf.ones(shape)], 3), axis=3)
+
+    # Mask the pixels if bordering artifacts appear
+    valid_mask = tf.ones_like(keypoint_map) if valid_mask is None else valid_mask
+    valid_mask = tf.to_float(valid_mask[..., tf.newaxis])  # for GPU
+    valid_mask = tf.space_to_depth(valid_mask, config['grid_size'])
+    valid_mask = tf.reduce_prod(valid_mask, axis=3)  # AND along the channel dim
+
+    loss = tf.losses.sparse_softmax_cross_entropy(labels=labels, logits=logits)
+    loss = tf.reduce_sum(valid_mask * loss) / tf.reduce_sum(valid_mask)
+    loss /= tf.to_float(tf.shape(logits)[-1])
```

Comment:
Why do you need to divide by this?
The previous line `loss = tf.reduce_sum(valid_mask * loss) / tf.reduce_sum(valid_mask)` should already have computed the right mean, shouldn't it?